### PR TITLE
Refactor qConstrainedLogEI

### PR DESCRIPTION
**Pull request by sawa3030**
## Motivation
- Address the refactoring suggestions raised in https://github.com/optuna/optuna/pull/6744#pullrequestreview-4783616664.
- Follow the draft refactoring proposed in https://github.com/optuna/optuna/pull/6744#issuecomment-5113806383.

## Description of the changes
- Simplify the handling of QMC seeds.
- Avoid accessing private attributes from `qConstrainedLogEI`.

### CODE REVIEW

**Comment by sawa3030** [comment]
I’ll mark this PR as ready for review once #6744 has been merged.

**Review by nabenabe0928 — APPROVED** [review]
LGTM

**Comment by not522** [comment]
@kAIto47802 Could you review this PR?

File: optuna/_gp/acqf.py [hunk 1]
```diff
@@ -15,10 +15,16 @@
...
 if TYPE_CHECKING:
+    from typing import Protocol
+
     import torch
 
     from optuna._gp.gp import GPRegressor
     from optuna._gp.search_space import SearchSpace
+
+    class PerSampleLogUtilityType(Protocol):
+        def compute_per_sample_log_utilility(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (kAIto47802) [inline optuna/_gp/acqf.py]: typo
```suggestion
        def compute_per_sample_log_utility(self, x: torch.Tensor) -> torch.Tensor:
```

File: optuna/_gp/acqf.py [hunk 2]
```diff
@@ -155,16 +161,20 @@ def __init__(
             qmc_seed=qmc_seed,
             stabilizing_noise=stabilizing_noise,
         )
+        n_running = len(normalized_params_of_running_trials)
+        self._per_sample_log_util_shape = (n_qmc_samples, n_running + 1)
         super().__init__(gpr.length_scales, search_space)
 
-    def eval_acqf(self, x: torch.Tensor) -> torch.Tensor:
+    def compute_per_sample_log_utilility(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (kAIto47802) [inline optuna/_gp/acqf.py]:
```suggestion
    def compute_per_sample_log_utility(self, x: torch.Tensor) -> torch.Tensor:
```

File: optuna/_gp/acqf.py [hunk 3]
```diff
@@ -155,16 +161,20 @@ def __init__(
...
+            return_tensor_shape = x.shape[:-1] + self._per_sample_log_util_shape
+            return torch.zeros(return_tensor_shape, dtype=torch.float64)
         # NOTE(nabenabe): See Eq. (10) of https://arxiv.org/pdf/2310.20708
         y_post = self._cond_gpr.sample_joint_posterior(x)
-        log_improvement = (y_post - self._threshold).clamp_min_(_EPS).log()
-        return _compute_mean_max_utility(log_improvement)
+        return (y_post - self._threshold).clamp_min_(_EPS).log()
+
+    def eval_acqf(self, x: torch.Tensor) -> torch.Tensor:
+        return _compute_mean_max_utility(self.compute_per_sample_log_utilility(x))
```
Comment (kAIto47802) [inline optuna/_gp/acqf.py]:
```suggestion
        return _compute_mean_max_utility(self.compute_per_sample_log_utility(x))
```

File: optuna/_gp/acqf.py [hunk 4]
```diff
@@ -232,10 +242,12 @@ def __init__(
         )
         super().__init__(gpr.length_scales, search_space)
 
-    def eval_acqf(self, x: torch.Tensor) -> torch.Tensor:
+    def compute_per_sample_log_utilility(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (kAIto47802) [inline optuna/_gp/acqf.py]:
```suggestion
    def compute_per_sample_log_utility(self, x: torch.Tensor) -> torch.Tensor:
```

File: optuna/_gp/acqf.py [hunk 5]
```diff
@@ -232,10 +242,12 @@ def __init__(
...
         y_post = self._cond_gpr.sample_joint_posterior(x)
-        log_prob = torch.nn.functional.logsigmoid((y_post - self._threshold) / self._tau)
-        return _compute_mean_max_utility(log_prob)
+        return torch.nn.functional.logsigmoid((y_post - self._threshold) / self._tau)
+
+    def eval_acqf(self, x: torch.Tensor) -> torch.Tensor:
+        return _compute_mean_max_utility(self.compute_per_sample_log_utilility(x))
```
Comment (kAIto47802) [inline optuna/_gp/acqf.py]:
```suggestion
        return _compute_mean_max_utility(self.compute_per_sample_log_utility(x))
```

File: optuna/_gp/acqf.py [hunk 6]
```diff
@@ -320,48 +332,38 @@ def __init__(
...
         super().__init__(gpr.length_scales, search_space)
 
     def eval_acqf(self, x: torch.Tensor) -> torch.Tensor:
-        y_post = self._acqf._cond_gpr.sample_joint_posterior(x)
-        log_feasible_improvement = (
-            (y_post - self._acqf._threshold).clamp_min_(_EPS).log()
-            if not np.isneginf(self._acqf._threshold)
-            else torch.zeros_like(y_post, dtype=torch.float64)
+        log_feasible_improvement = self._acqf.compute_per_sample_log_utilility(x) + sum(
+            acqf.compute_per_sample_log_utilility(x) for acqf in self._constraints_acqf_list
```
Comment (kAIto47802) [inline optuna/_gp/acqf.py]:
```suggestion
        log_feasible_improvement = self._acqf.compute_per_sample_log_utility(x) + sum(
            acqf.compute_per_sample_log_utility(x) for acqf in self._constraints_acqf_list
```

**Review by kAIto47802 — CHANGES_REQUESTED** [review]

**Comment by sawa3030** [comment]
@kAIto47802 Thank you for pointing it out. I’ve applied the change!

File: optuna/samplers/_gp/sampler.py [hunk 7]
```diff
@@ -50,6 +50,10 @@
 _logger = logging.getLogger(__name__)
 
 EPS = 1e-10
+# NOTE(nabe): Do NOT use 1 << 31 to avoid overflow. Multi-objective and constrained optimizations
+# use qmc_seed + buffer. Since buffer is sufficiently small (like ~30), using `1 << 30` guarantees
+# no overflow.
```
Comment (kAIto47802) [inline optuna/samplers/_gp/sampler.py]: This comment seems inaccurate. `SobolEngine` internally uses `torch.Generator.manual_seed`, whose officially documented valid seed range is [-2^63, 2^64 - 1]. Therefore, 1 << 31 is within the supported range:
https://docs.pytorch.org/docs/2.13/generated/torch.Generator.html#torch.Generator.manual_seed

Also, the scalar result of `RandomState.randint(...)` and the subsequent i + 1 addition are handled as Python integers, so they do not overflow at the 32-bit boundary.

```suggestion
# NOTE(nabe): Multi-objective and constrained optimization may derive QMC seeds by adding small
# (e.g., around 30) offsets to a base seed. Use `1 << 30` as a conservative upper bound.
```

**Comment by kAIto47802** [comment]
Thank you for the update! It almost looks good to me, with some comments.

**Comment by sawa3030** [comment]
Thank you! I had missed that point. I’ve updated the comment accordingly. PTAL!

File: optuna/samplers/_gp/sampler.py [hunk 8]
```diff
@@ -450,6 +453,12 @@ def _sample_relative_impl(
             )
         self._gprs_cache_list = gprs_list
 
+        print(_MAX_QMC_SEED_VALUE)
+        print(self._rng.rng.randint(_MAX_QMC_SEED_VALUE))
+        print(self._rng.rng.randint(_MAX_QMC_SEED_VALUE))
+        print(self._rng.rng.randint(_MAX_QMC_SEED_VALUE))
+        print(self._rng.rng.randint(_MAX_QMC_SEED_VALUE))
+
```
Comment (kAIto47802) [inline optuna/samplers/_gp/sampler.py]: Thank you for the update.
But, the latest commit mixes unnesessary debug prints.
```suggestion
```

**Review by kAIto47802 — CHANGES_REQUESTED** [review]

File: optuna/samplers/_gp/sampler.py (hunk 8 above)
Comment (sawa3030) [reply optuna/samplers/_gp/sampler.py]: Sorry for that! Updated!

**Review by kAIto47802 — APPROVED** [review]
Thank you for the update! LGTM
